### PR TITLE
fuzz: replace every fuzzer-controlled while loop with a macro

### PR DESCRIPTION
**Pull request by apoelstra**
Limits the number of iterations to 1000 rather than letting the fuzzer do millions or billions of iterations on a single core.

### CODE REVIEW

**Comment by apoelstra** [comment]
There are several loops in the fuzz test where the number of iterations is directly controlled by the fuzzer, which oftentimes will go for millions of iterations, leading to certain fuzz tests taking forever on a single core. This PR limits these fuzzer-controlled loops to 1000 iterations, an arbitrary high number.

IMHO letting the fuzzer control the number of iterations directly like this, with no restrictions, is a bad idea because the fuzzer will naturally try to hit super high numbers.

I'm looking for a concept ACK (should we limit these) and an approach ACK (is doing this with a macro a reasonable way to go?). Actual ACKs/merges would also be nice :).

**Comment by apoelstra** [comment]
On an EPYC 7502 with 256Gb of RAM:

before, -j1
```
real    67m58.883s
user    152m7.845s
sys     4m36.080s
```

after, -j1
```
real    35m57.697s
user    103m10.803s
sys     3m52.910s
```

before, -j32
```
real    60m6.374s
user    153m29.631s
sys     5m14.991s
```

after, -j32
```
real    28m54.078s
user    102m52.337s
sys     4m16.605s
```

**Comment by MarcoFalke** [comment]
Concept ACK. @DrahtBot please list the conflicts for the alternative solution that has been proposed.

**Comment by MarcoFalke** [comment]
Ok, that didn't work. Basically, it would be nice to be able to still overwrite the limit per fuzz-case:

* #22454: 300
* #22453: 3000
* #22444: 3000
* ...

**Comment by apoelstra** [comment]
@MarcoFalke I added a #define variable and capped the `tx_pool` test to 300 iterations to demonstrate how to use it.

File: src/test/fuzz/tx_pool.cpp [hunk 1]
```diff
@@ -142,7 +144,7 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
         return c.out.nValue;
     };
 
-    while (fuzzed_data_provider.ConsumeBool()) {
+    FUZZER_CONTROLLED_LOOP(fuzzed_data_provider) {
```
Comment (MarcoFalke) [inline src/test/fuzz/tx_pool.cpp:147]: it would be nice to be able to pass the limit to the loop (in the same line). Otherwise it will get messy if there are nested loops and one forgets to adjust the limit or the loops are moved without moving the limit, which is in a different line.

Comment (apoelstra) [reply src/test/fuzz/tx_pool.cpp]: Sure -- replaced the extra #define variable with a new macro argument

Comment (apoelstra) [reply src/test/fuzz/tx_pool.cpp]: Maybe we should just drop the "default 1000" one and always explicitly set the maximum number of iterations?

File: src/test/fuzz/addrman.cpp [hunk 2]
```diff
@@ -44,7 +44,7 @@ FUZZ_TARGET_INIT(addrman, initialize_addrman)
             addr_man.m_asmap.clear();
         }
     }
-    while (fuzzed_data_provider.ConsumeBool()) {
+    FUZZER_CONTROLLED_LOOP(fuzzed_data_provider) {
```
Comment (MarcoFalke) [inline src/test/fuzz/addrman.cpp:47]:
```suggestion
    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool) {
```

Since some loops (prefer to?) use `fuzzed_data_provider.remaining_bytes()`, maybe make this configurable as well? Feel free to ignore the naming suggestion.

File: src/test/fuzz/tx_pool.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/test/fuzz/tx_pool.cpp]: Yeah, I wouldn't mind that. An additional `, 1000` doesn't seem too verbose.

**Comment by practicalswift** [comment]
Concept ACK

Nice idea!

File: src/test/fuzz/addrman.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/test/fuzz/addrman.cpp:47]: @apoelstra Are you still working on this? If not, I am happy to pick up

Comment (apoelstra) [reply src/test/fuzz/addrman.cpp:47]: Hi, sorry, I can pick this up now.

**Comment by apoelstra** [comment]
Redid the PR. Now it takes the macro from #22649 and just applies it to every single remaining fuzzer loop.

**Comment by practicalswift** [comment]
Thanks for fixing this!

Is the mixing of `LIMITED_WHILE (` and `LIMITED_WHILE(` intentional? (Personally I prefer the latter.)

ACK modulo spacing consistency :)

**Comment by laanwj** [comment]
> There are several loops in the fuzz test where the number of iterations is directly controlled by the fuzzer, which oftentimes will go for millions of iterations, leading to certain fuzz tests taking forever on a single core. 

Concept and code review ACK be655e8b8111f3a7cd7909203c2bac3d16751406

I would slightly prefer this "arbitrary high number" to be a constant as it's repeated everywhere.

**Comment by MarcoFalke** [comment]
> I would slightly prefer this "arbitrary high number" to be a constant as it's repeated everywhere.

I'd say the number is dependent on the test context. The current ones in master are 30, 300, 3000, and 30000.

I think each number needs to be picked carefully to be in the right 10^x.

> Is the mixing of LIMITED_WHILE ( and LIMITED_WHILE( intentional? (Personally I prefer the latter.)

clang-format(-diff) should take care of those in any case.

**Comment by apoelstra** [comment]
Oops, I will fix the spacing inconsistency. It probably happened when I switched from manual edits to using sed.

@MarcoFalke I think we could probably start with an "anbitrary high number", say 30000 (or 10000), and then when when tests reveal that certain tests are taking too long we can reduce them in individual cases. The current situation effectively has a limit of infinity and things are mostly OK, so I don' t think this'll be a big deal.

Once we're at 5-digit numbers I don't think there are any situations where we'd want to *increase* them.

**Comment by MarcoFalke** [comment]
Sorry, will review soon

File: src/test/fuzz/versionbits.cpp [hunk 3]
```diff
@@ -199,7 +199,7 @@ FUZZ_TARGET_INIT(versionbits, initialize)
     const uint32_t signalling_mask = fuzzed_data_provider.ConsumeIntegral<uint32_t>();
 
     // mine prior periods
-    while (fuzzed_data_provider.remaining_bytes() > 0) {
+    LIMITED_WHILE(fuzzed_data_provider.remaining_bytes() > 0, 10000) {
```
Comment (MarcoFalke) [inline src/test/fuzz/versionbits.cpp:202]: Why is this needed? There is already a `break` below, no?

File: src/test/fuzz/rpc.cpp [hunk 4]
```diff
@@ -348,7 +348,7 @@ FUZZ_TARGET_INIT(rpc, initialize_rpc)
         return;
     }
     std::vector<std::string> arguments;
-    while (fuzzed_data_provider.ConsumeBool()) {
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10000) {
```
Comment (MarcoFalke) [inline src/test/fuzz/rpc.cpp:351]: This seems a bit high? `ConsumeRPCArgument` might call `ConsumeArrayRPCArgument`, so the limit is 100'000'000. A user shouldn't be passing in more than 100 items. So any value between 100-1000 seems more appropriate.

See also https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40648

Which is running base58 encoding (expensive) of one rpc arg 6000 times, thus times out. Your limit of 10000 wouldn't affect that, I think.

**Review by maflcko — APPROVED** [review]
ACK, beside comments

File: src/test/fuzz/versionbits.cpp (hunk 3 above)
Comment (apoelstra) [reply src/test/fuzz/versionbits.cpp:202]: Yep, you're right, I misread the loop's exit condition.

File: src/test/fuzz/rpc.cpp (hunk 4 above)
Comment (apoelstra) [reply src/test/fuzz/rpc.cpp:351]: ACK. I reduced it to 100.

I worry that reducing this will meaningfully affect test coverage since we presumably want to test what happens when you give gazillions of RPC arguments ... but probably a general-purpose fuzz test is the wrong place to try to hit that.

Comment (MarcoFalke) [reply src/test/fuzz/rpc.cpp]: Sorry, I also meant in the line above in this file. 100 * 100 is still 10k.

File: src/test/fuzz/versionbits.cpp [hunk 5]
```diff
@@ -199,7 +199,7 @@ FUZZ_TARGET_INIT(versionbits, initialize)
     const uint32_t signalling_mask = fuzzed_data_provider.ConsumeIntegral<uint32_t>();
 
     // mine prior periods
-    while (fuzzed_data_provider.remaining_bytes() > 0) {
+    while(fuzzed_data_provider.remaining_bytes() > 0) { // early exit; no need for LIMITED_WHILE
```
Comment (MarcoFalke) [inline src/test/fuzz/versionbits.cpp:202]:
```suggestion
    while (fuzzed_data_provider.remaining_bytes() > 0) { // early exit; no need for LIMITED_WHILE
```

nit

Comment (apoelstra) [reply src/test/fuzz/versionbits.cpp:202]: fixed

File: src/test/fuzz/rpc.cpp [hunk 6]
```diff
@@ -294,7 +294,7 @@ std::string ConsumeScalarRPCArgument(FuzzedDataProvider& fuzzed_data_provider)
 std::string ConsumeArrayRPCArgument(FuzzedDataProvider& fuzzed_data_provider)
 {
     std::vector<std::string> scalar_arguments;
-    while (fuzzed_data_provider.ConsumeBool()) {
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10000) {
```
Comment (MarcoFalke) [inline src/test/fuzz/rpc.cpp:297]: This one I meant, too ;)

Comment (MarcoFalke) [reply src/test/fuzz/rpc.cpp:297]: https://github.com/bitcoin/bitcoin/pull/22508#discussion_r748396777

**Comment by MarcoFalke** [comment]
Some times:

master | this | relative | name
```
	206.2	76.67	0.3718234724	rpc			
	149.06	157.45	1.056286059	process_messages																				
	27.21	27.19	0.9992649761	addrman																				
	24.19	24.89	1.028937578	script_sign																				
	23.93	26.08	1.089845382	process_message																				
	23.32	20.44	0.8765008576	coins_view																				
	15.22	15.75	1.034822602	policy_estimator																				
	14.46	15.32	1.059474412	script_ops																				
	13.06	12.85	0.9839203675	bloom_filter																				
	12.35	12.41	1.0048583	pow																				
	11.93	8.79	0.7367979883	connman																				
	10.71	9.38	0.8758169935	net																				
	9.22	7.21	0.7819956616	node_eviction																				
	7.7	8.23	1.068831169	rbf

**Comment by MarcoFalke** [comment]
So yeah, this should fix https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40648

**Comment by MarcoFalke** [comment]
cr ACK 214d9055acdd72189a2f415477ce472ca8db4191
